WalletWasabi/Extensions/Extensions.cs

```diff
@@ -0,0 +1,20 @@
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Interfaces;
+
+namespace WalletWasabi.Extensions
+{
+	public static class Extensions
+	{
+		/// <summary>
+		/// Waits until all previously enqueued tasks have been finished. Note: Queue doesn't necesarilly has to be empty afterwards.
+		/// </summary>
+		public static async Task WaitAsync(this IBackgroundTaskQueue queue, CancellationToken cancellationToken = default)
+		{
+			using var processed = new SemaphoreSlim(0);
+			var enqueue = queue.QueueBackgroundWorkItemAsync(_ => { processed.Release(); return ValueTask.CompletedTask; });
+			await Task.WhenAny(enqueue.AsTask(), Task.Delay(Timeout.Infinite, cancellationToken)).ConfigureAwait(false);
+			await processed.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
for this use case, what's the difference between `SemaphoreSlim` and `TaskCompletionSource`?

Reply:
well, one doesn't put a task into the queue, one puts synchronous callback which returns a task so I don't really have a task to await do I?

I don't know this is mostly for tests anyway. You shouldn't really wait for the queue in the production code. TaskCompletionSource seems like a low level thing 🤷‍♂️

Reply:
or maybe it can be replaced as a more efficient, I'll try that, just to learn 👀

Reply:
you can trigger it from a sync callback: https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcompletionsource.setresult?view=net-6.0#System_Threading_Tasks_TaskCompletionSource_SetResult

but anyway like you said it doesn't matter, i was just curious if there's a reason